src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -112,142 +75,94 @@ public static bool TryMoveDown(Project project, IProjectTree projectTree)
-            Requires.NotNull(nodes, nameof(nodes));
-            Requires.NotNull(target, nameof(target));
-
-            // Move added nodes to the top. 
-            var nodeToAddAbove = GetFirstChild(target);
-
-            // This could have been an empty folder and nodeToAddAbove is null, therefore
-            //     get the first child of the target's parent.
-            // Empty folders do not have a valid display order.
-            // In the future, if we want to order empty folders, meaning they have a valid display order, 
-            //     this will need to be extended to handle that behavior.
+            var includes = GetEvaluatedIncludes(projectTree);
 
-            // This recursively tries to find a node to work with. If there is none at all, this will break on project root.
-            var currentTarget = target;
-            while(nodeToAddAbove == null && currentTarget.IsFolder && !GetChildren(currentTarget).Any())
-            {
-                currentTarget = currentTarget.Parent;
-                nodeToAddAbove = GetFirstChild(currentTarget);
-            }
+            var elements = new List<ProjectItemElement>();
 
-            if (nodeToAddAbove != null && nodes.Any())
+            foreach (var include in includes)
             {
-                var didAllMoveSuccessfully = true;
+                // GetItemsByEvaluatedInclude is efficient and uses a MultiDictionary underneath.
+                //     It uses this: new MultiDictionary<string, ProjectItem>(StringComparer.OrdinalIgnoreCase);
+                var item = project.GetItemsByEvaluatedInclude(include).FirstOrDefault();
 
-                foreach (var node in nodes)
+                // We only care about adding one item associated with the evaluated include.
+                if (item?.Xml is ProjectItemElement element)
                 {
-                    if (!TryMoveAbove(project, node, nodeToAddAbove))
-                    {
-                        didAllMoveSuccessfully = false;
-                    }
+                    elements.Add(element);
                 }
-
-                return didAllMoveSuccessfully;
             }
 
-            return false;
+            return elements.ToImmutableArray();
         }
 
         /// <summary>
-        /// This moves the nodes above the target's first child. 
-        /// If there are no children with a valid display order of the target or the target is an empty folder, it will try the parent target's first child.
+        /// If the project gets any additional items from the callback, this will return those added items.
         /// </summary>
-        public static Task<bool> TryMoveNodesToTopAsync(ConfiguredProject configuredProject, IEnumerable<IProjectTree> nodes, IProjectTree target)
+        public static async Task<ImmutableArray<ProjectItemElement>> AddItems(ConfiguredProject configuredProject, IProjectLockService projectLockService, Func<Task> addItems)
```

Comment:
While the main usage of this is within a greater write lock, this function on its own has issues with taking a releasing locks. Between the 2 read locks, calling `addItems`, anyone else could have taken a write lock and changed the project. You would then try to move those items. This function will need to perform all of these steps within one lock:

1. Collecting previous items
2. Calling `addItems`
3. Collecting new items
4. **\*new\*** processing new items

```CSharp
using (var access = await projectLockService.UpgradeableReadLockAsync(ProjectLockFlags.StickyWrite))
{
  // collect previous items
  await addItems();
  if (!projectLockService.IsWriteLockHeld)
  {
    // addItems did not add anything, abort early.
    // this is optional. You can also condition calling the process on if any new items are found.
    return;
  }
  
  // will need to get the project from the access object
  project.ReevaluateIfNecessary();

  // collect new items
  await processNewItems(newItems); // new callback passed into here.
}
```

Reply:
Ok, I'll see what I can do here.